tools/fix_bss.py

```diff
@@ -0,0 +1,649 @@
+    generate_make_log,
+    find_compiler_command_line,
+    run_cfe,
+    SymbolTableEntry,
+    UcodeOp,
+)
+
+import elftools.elf.elffile
+import mapfile_parser.mapfile
+
+
+def read_u32(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(4), "big")
+
+
+def read_u16(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(2), "big")
+
+
+def read_s16(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(2), "big", signed=True)
+
+
+def fail(message: str):
+    print(f"{colorama.Fore.RED}{message}{colorama.Fore.RESET}", file=sys.stderr)
+    sys.exit(1)
+
+
+@dataclass
+class Reloc:
+    name: str
+    offset_32: int | None
+    offset_hi16: int | None
+    offset_lo16: int | None
+    addend: int
+
+
+@dataclass
+class Pointer:
+    name: str
+    addend: int
+    base_value: int
+    build_value: int
+
+
+# Read relocations from an ELF file section
+def read_relocs(object_path: Path, section_name: str) -> list[Reloc]:
+    with open(object_path, "rb") as f:
+        elffile = elftools.elf.elffile.ELFFile(f)
+        symtab = elffile.get_section_by_name(".symtab")
+        data = elffile.get_section_by_name(section_name).data()
+
+        reloc_section = elffile.get_section_by_name(f".rel{section_name}")
+        if reloc_section is None:
+            return []
+
+        relocs = []
+        offset_hi16 = 0
+        for reloc in reloc_section.iter_relocations():
+            reloc_offset = reloc.entry["r_offset"]
+            reloc_type = reloc.entry["r_info_type"]
+            reloc_name = symtab.get_symbol(reloc.entry["r_info_sym"]).name
+
+            if reloc_type == 2:  # R_MIPS_32
+                offset_32 = reloc_offset
+                addend = int.from_bytes(
+                    data[reloc_offset : reloc_offset + 4], "big", signed=True
+                )
+                relocs.append(Reloc(reloc_name, offset_32, None, None, addend))
+            elif reloc_type == 4:  # R_MIPS_26
+                pass
+            elif reloc_type == 5:  # R_MIPS_HI16
+                offset_hi16 = reloc_offset
+            elif reloc_type == 6:  # R_MIPS_LO16
+                offset_lo16 = reloc_offset
+                addend_hi16 = int.from_bytes(
+                    data[offset_hi16 + 2 : offset_hi16 + 4], "big", signed=False
+                )
+                addend_lo16 = int.from_bytes(
+                    data[offset_lo16 + 2 : offset_lo16 + 4], "big", signed=True
+                )
+                addend = (addend_hi16 << 16) + addend_lo16
+                relocs.append(Reloc(reloc_name, None, offset_hi16, offset_lo16, addend))
+            else:
+                raise NotImplementedError(f"Unsupported relocation type: {reloc_type}")
+
+        return relocs
+
+
+def get_file_pointers(
+    file: mapfile_parser.mapfile.File,
+    base: BinaryIO,
+    build: BinaryIO,
+) -> list[Pointer]:
+    pointers = list[Pointer]()
```

Comment:
Just so we don't forget: the `pointers = list[Pointer]()` syntax probably doesn't work on old python versions (used several times now)